lib/charms/mysql/v0/mysql.py

```diff
@@ -311,6 +334,256 @@ class RouterUser:
+        """Application peer relation data object."""
+        if self.peers is None:
+            return {}
+
+        return self.peers.data[self.app]
+
+    @property
+    def unit_peer_data(self) -> Dict:
+        """Unit peer relation data object."""
+        if self.peers is None:
+            return {}
+
+        return self.peers.data[self.unit]
+
+    @property
+    def _is_peer_data_set(self):
+        return (
+            self.app_peer_data.get("cluster-name")
+            and self.get_secret("app", ROOT_PASSWORD_KEY)
+            and self.get_secret("app", SERVER_CONFIG_PASSWORD_KEY)
+            and self.get_secret("app", CLUSTER_ADMIN_PASSWORD_KEY)
+            and self.get_secret("app", MONITORING_PASSWORD_KEY)
+            and self.get_secret("app", BACKUPS_PASSWORD_KEY)
+        )
+
+    def _get_secret_from_juju(self, scope: str, key: str) -> Optional[str]:
+        """Retrieve and return the secret from the juju secret storage."""
+        if scope == "unit":
+            secret_id = self.unit_peer_data.get(SECRET_ID_KEY)
+
+            if not self.unit_secrets and not secret_id:
+                logger.debug("Getting a secret when no secrets added in juju")
+                return None
+
+            if not self.unit_secrets:
+                secret = self.model.get_secret(id=secret_id)
+                content = secret.get_content()
+                self.unit_secrets = content
+
+            logger.debug(f"Retrieved secret {key} for unit")
+            return self.unit_secrets.get(key)
+
+        secret_id = self.app_peer_data.get(SECRET_ID_KEY)
+
+        if not self.app_secrets and not secret_id:
+            logger.debug("Getting a secret when no secrets added in juju")
+            return None
+
+        if not self.app_secrets:
+            secret = self.model.get_secret(id=secret_id)
+            content = secret.get_content()
+            self.app_secrets = content
+
+        logger.debug(f"Retrieved secret {key} for app")
+        return self.app_secrets.get(key)
+
+    def _get_secret_from_databag(self, scope: str, key: str) -> Optional[str]:
+        """Retrieve and return the secret from the peer relation databag."""
+        if scope == "unit":
+            return self.unit_peer_data.get(key)
+
+        return self.app_peer_data.get(key)
+
+    def get_secret(self, scope: str, key: str, fallback_key: str = None) -> Optional[str]:
+        """Get secret from the secret storage.
+
+        Retrieve secret from juju secrets backend if secret exists there.
+        Else retrieve from peer databag (with key or fallback_key). This is to
+        account for cases where secrets are stored in peer databag but the charm
+        is then refreshed to a newer revision.
+        """
+        if scope not in ["unit", "app"]:
+            raise MySQLSecretError(f"Invalid secret scope: {scope}")
+
+        if ops.jujuversion.JujuVersion.from_environ().has_secrets:
+            return self._get_secret_from_juju(scope, key)
```

Comment:
If the charm is running on Juju 3.1.x (no Juju upgrade) and it was updated from the old charm version to this version. It still has the password in the databag, but this function returns them from an empty secret.
We discussed this case on DS 😞 
please ask @dragomirp how we handle this case.

Reply:
addressed in b3f0e02aa67f1ebf1a018e27a104be32c0decb3d